### ISSUE TITLE
OOM when using events

### ISSUE THREAD

**Issue by pelikhan** [microsoft/pxt-microbit#2727]
If you button smash enough A, this program will OOM.
```
input.onButtonPressed(Button.A, function () {
    basic.showString("OOMO OOM OMOOOMMO MOOM OMOMOMOO")
})
```

**Comment by finneyj** [comment]
Thanks all.

@pelikhan Is this build configured for the new Javascript style event concurrency? i.e. what have you go this set to in this build of PXT?

https://github.com/lancaster-university/microbit-dal/blob/js-event-semantics/inc/core/MicroBitConfig.h#L204

This new behaviour probably is heavier on RAM, and uses a new locking primitive for condition synchronization (https://github.com/lancaster-university/microbit-dal/commit/741335200d292161a69aee86c1ed92f5dd59816a).

I can't see any other changes that might cause this though... let me know.

**Comment by finneyj** [comment]
Just performed a quick test in C++, with both settings of the above concurrency mode (DAL style and JS style) and there is a marked difference in RAM consumption.

With the following code:

```
void onButtonA(MicroBitEvent) {
    uBit.display.scroll("HELLO WORLD!!!");
}

void onButtonB(MicroBitEvent) {
    microbit_heap_print();
}

int main() {
    uBit.init();
    uBit.serial.printf("\n\n *** START ***\n");

    uBit.messageBus.listen(MICROBIT_ID_BUTTON_A, MICROBIT_BUTTON_EVT_CLICK, onButtonA);
    uBit.messageBus.listen(MICROBIT_ID_BUTTON_B, MICROBIT_BUTTON_EVT_CLICK, onButtonB);

    release_fiber();
    return 0;
}
```

A five fast clicks of BUTTON_A followed by BUTTON_B, we get the following amounts of RAM available:

MESSAGE_BUS_CONCURRENT_LISTENERS: 796+1024 bytes (1820 bytes free)
MESSAGE_BUS_CONCURRENT_EVENTS: 164+696 bytes (860 bytes free)

in other words, JS semantics are burning fiber contexts... its possible we can't afford this behaviour on micro:bit, unless we either find more RAM (unlikely) or decide to make event handlers fully non-reentrant...

Pretty sure this isn't the behaviour that @tballmsft was looking for, but is perhaps a valid compromise?
